# Task 0: Baseline Model Profiling

## Metrics Profiled:
- Model Size (MB)
- Peak Memory Usage (MB)
- Average Memory Usage (MB)
- Inference Latency (ms per batch)
- Energy Consumption (mJ)
- MACs (Multiply-Accumulate Operations)
- Top-1 and Top-5 Accuracy (Train and Test)

In [22]:
# Setup for Colab and Local environments
import sys
import os

# Check if running on Colab
try:
    import google.colab # type: ignore
    IN_COLAB = True
    print("Running on Google Colab")
    
    # Mount Google Drive
    from google.colab import drive # type: ignore
    drive.mount('/content/drive')
    
    # Add project directory to Python path
    # Change this path to match your Google Drive folder structure
    PROJECT_PATH = '/content/drive/MyDrive/AI624-Edge-Devices/Assignment-1.1'
    sys.path.append(PROJECT_PATH)
    print(f"Added {PROJECT_PATH} to Python path")
    
except ImportError:
    IN_COLAB = False
    print("Running locally")

# Enable PyTorch MPS fallback (macOS only)
from utils import enablePytorchFallback
enablePytorchFallback()

Running locally
✓ MPS fallback enabled: Unsupported operations will run on CPU


In [10]:
# Import required modules
import os
import torch
from utils import VGG16_CIFAR, getDeviceFromHardware, get_base_path

In [11]:
# Auto-detect device
DEVICE = getDeviceFromHardware()
print(f"Using device: {DEVICE}")

# Set base path (auto-detects local vs Colab)
BASE_PATH = get_base_path()
print(f"Base path: {BASE_PATH}")

Using device: mps
Base path: ./


In [12]:
# Create CIFAR-10 model instance
print("="*80)
print("CIFAR-10 MODEL SETUP")
print("="*80)

cifar10_model = VGG16_CIFAR(num_classes=10, is_pruned=False)
print("✓ CIFAR-10 model architecture created")

CIFAR-10 MODEL SETUP
✓ CIFAR-10 model architecture created


In [13]:
# Load pre-trained weights
cifar10_model_path = os.path.join(BASE_PATH, 'models/cifar10_vgg16.pt')
cifar10_model.load(cifar10_model_path, device=DEVICE)
print("✓ CIFAR-10 model weights loaded")

Loading model from ./models/cifar10_vgg16.pt
Model loaded successfully
✓ CIFAR-10 model weights loaded


In [14]:
# Get datasets
cifar10_train_dataset, cifar10_test_dataset = cifar10_model.get_train_test_split(base_path=BASE_PATH)
print(f"✓ CIFAR-10 datasets loaded")
print(f"  Train size: {len(cifar10_train_dataset)}")
print(f"  Test size: {len(cifar10_test_dataset)}")

Dataset found at ./datasets/cifar-10-batches-py
✓ CIFAR-10 datasets loaded
  Train size: 50000
  Test size: 10000


In [15]:
# Create dataloaders
cifar10_train_loader, cifar10_test_loader = cifar10_model.get_data_loaders(
    cifar10_train_dataset, 
    cifar10_test_dataset, 
    batch_size=128
)
print("✓ CIFAR-10 dataloaders created")
print(f"  Train batches: {len(cifar10_train_loader)}")
print(f"  Test batches: {len(cifar10_test_loader)}")

✓ CIFAR-10 dataloaders created
  Train batches: 391
  Test batches: 79


In [16]:
# Execute comprehensive profiling for CIFAR-10
cifar10_model.execute_task0(cifar10_train_loader, cifar10_test_loader, device=DEVICE)


TASK 0: BASELINE MODEL PROFILING RESULTS

--- CIFAR-10 VGG16-BN ---
Model Size:          58.25 MB
Peak Memory:         0.00 MB
Average Memory:      0.00 MB
Latency (per batch): 12.24 ms
Energy:              Not available (Linux with RAPL required)
MACs:                314,002,944

Evaluating test accuracy...
Test Top-1:          94.16%
Test Top-5:          99.71%

Evaluating train accuracy...
Train Top-1:         100.00%
Train Top-5:         100.00%



In [17]:
# Create CIFAR-100 model instance
print("="*80)
print("CIFAR-100 MODEL SETUP")
print("="*80)

cifar100_model = VGG16_CIFAR(num_classes=100, is_pruned=False)
print("✓ CIFAR-100 model architecture created")

CIFAR-100 MODEL SETUP
✓ CIFAR-100 model architecture created


In [18]:
# Load pre-trained weights
cifar100_model_path = os.path.join(BASE_PATH, 'models/cifar100_vgg16.pt')
cifar100_model.load(cifar100_model_path, device=DEVICE)
print("✓ CIFAR-100 model weights loaded")

Loading model from ./models/cifar100_vgg16.pt
Model loaded successfully
✓ CIFAR-100 model weights loaded


In [19]:
# Get datasets
cifar100_train_dataset, cifar100_test_dataset = cifar100_model.get_train_test_split(base_path=BASE_PATH)
print(f"✓ CIFAR-100 datasets loaded")
print(f"  Train size: {len(cifar100_train_dataset)}")
print(f"  Test size: {len(cifar100_test_dataset)}")

Dataset found at ./datasets/cifar-100-python
✓ CIFAR-100 datasets loaded
  Train size: 50000
  Test size: 10000


In [20]:
# Create dataloaders
cifar100_train_loader, cifar100_test_loader = cifar100_model.get_data_loaders(
    cifar100_train_dataset, 
    cifar100_test_dataset, 
    batch_size=128
)
print("✓ CIFAR-100 dataloaders created")
print(f"  Train batches: {len(cifar100_train_loader)}")
print(f"  Test batches: {len(cifar100_test_loader)}")

✓ CIFAR-100 dataloaders created
  Train batches: 391
  Test batches: 79


In [21]:
# Execute comprehensive profiling for CIFAR-100
cifar100_model.execute_task0(cifar100_train_loader, cifar100_test_loader, device=DEVICE)


TASK 0: BASELINE MODEL PROFILING RESULTS

--- CIFAR-100 VGG16-BN ---
Model Size:          58.43 MB
Peak Memory:         0.00 MB
Average Memory:      0.00 MB
Latency (per batch): 6.60 ms
Energy:              Not available (Linux with RAPL required)
MACs:                314,049,024

Evaluating test accuracy...
Test Top-1:          74.00%
Test Top-5:          90.56%

Evaluating train accuracy...
Train Top-1:         99.93%
Train Top-5:         100.00%



## Task 0: Baseline Model Profiling Results

### Results Summary

| Metric                   | CIFAR-10    | CIFAR-100   |
| ------------------------ | ----------- | ----------- |
| Model Size (MB)          | 58.25       | 58.43       |
| Peak Memory (MB)         | 270.03      | 270.03      |
| Average Memory (MB)      | 78.47       | 78.48       |
| Latency (ms/batch)       | 12.24       | 6.60        |
| Energy Consumption (mJ)  | N/A         | N/A         |
| MACs                     | 314,002,944 | 314,049,024 |
| Test Top-1 Accuracy (%)  | 94.16       | 74.00       |
| Test Top-5 Accuracy (%)  | 99.71       | 90.56       |
| Train Top-1 Accuracy (%) | 100.00      | 99.93       |
| Train Top-5 Accuracy (%) | 100.00      | 100.00      |

### Implementation Details

- **Notebook**: `task0.ipynb`
- **Batch Size**: 128
- **Device Used**:
  - MPS (Apple Silicon) for latency and accuracy measurements
  - CPU for memory profiling (PyTorch profiler limitation)

### Important Notes

#### 1. Memory Profiling Limitation
**Memory usage profiling only works accurately on CPU devices.**

- PyTorch's profiler can only track CPU memory usage accurately
- When using CUDA or MPS, memory is allocated on GPU (not tracked by CPU profiler)
- Results show 0.00 MB on GPU devices

**Solution**: Run profiling with `DEVICE = 'cpu'` for memory metrics
- CIFAR-10 CPU results: Peak = 270.03 MB, Average = 78.47 MB
- CIFAR-100 CPU results: Peak = 270.03 MB, Average = 78.48 MB

#### 2. Energy Consumption
**Energy measurement not available (requires Linux with RAPL support).**

- pyJoules requires direct hardware access via Intel RAPL
- RAPL is not available on:
  - macOS (Apple Silicon or Intel)
  - Windows
  - Virtual machines (including Google Colab)
- Official documentation: https://github.com/powerapi-ng/pyJoules/blob/master/docs/index.rst#limitation

#### 3. Model Accuracy

- **CIFAR-10**: 94.16% test accuracy
- **CIFAR-100**: 74.00% test accuracy

#### 4. Computational Complexity
- Both models have similar MACs (~314M operations)
- Classifier difference (10 vs 100 classes) has minimal impact on total computation